### PR TITLE
Add fork selection RFC

### PR DESCRIPTION
**Pull request by aeyakovenko**
#### Problem

Lack of RFC on fork selection.

#### Summary of Changes

RFC on fork selection and simulation.

Fixes # #1983 (choice based on maximizing finality should be censorship resistant)

### CODE REVIEW

File: rfcs/0005-fork-selection.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
+# Fork Selection
+
+This RFC defines Solana's *Nakomoto Fork Selection* algorithm based on time locks. It satisfies the following properties:
+
+* A voter can eventually recover from voting on a fork that never sees majority votes.
+* If the voters share a common ancestor then they will converge to a fork containing that ancestor no matter how they are partitioned.  Although it may not be the latest possible ancestor at the start of the partition.
+* Rollback requires exponentially more time for older votes than for newer votes.
+* Voters can independently configure a vote threshold they would like to see before committing a vote to a higher lockout.  This allows each voter to make a trade-off of risk and reward.
+* Voter act greedy to maximize their rewards.  The greedy choice leads to convergance of the network.
```
Comment (rob-solana) [inline rfcs/0005-fork-selection.md]: Voters

File: rfcs/0005-fork-selection.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+For solana `time` uniquely identifies a specific leader for fork generation.  At any given time only 1 leader which can be computed from the ledger itself can propose a fork.  For more details see [fork generation](0002-fork-generation.md) and [leader rotation](0004-leader-rotation.md).
+
+## Algorithm
+
+The basic idea to this approach is to stack consensus votes.  Votes in the stack must be for forks that are descendent from each other, and for forks that are valid in the ledger they are submitted to.  Each consensus vote has a `lockout` in units of time before it can be discarded.  When a vote is added to the stack the lockouts of all the previous votes in the stack are doubled (see [Rollback](#Rollback)).  With each new vote a voter commits the previous votes to a ever increasing lockout.  Since at 32 votes we can consider the system to be at `max lockout` any votes with a lockout above `1<<32` are dequeued (fifo).  Dequeuing a vote is the trigger for a reward.  If a vote expires before it is dequeued, it and all the votes above it are popped (filo) from the vote stack.  The voter needs to start rebuilding the stack from that point.
```
Comment (rob-solana) [inline rfcs/0005-fork-selection.md]: "to an ever increasing"

File: rfcs/0005-fork-selection.md [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+For example, a vote stack with the following state:
+
+| vote | vote time | lockout | lock expiration time |
+|-----:|----------:|--------:|---------------------:|
+|    4 |         4 |      2  |                    6 |
+|    3 |         3 |      4  |                    7 |
+|    2 |         2 |      8  |                   10 |
+|    1 |         1 |      16 |                   17 |
+
+`Vote 5` is at time 9, and the resulting state is
```
[the remark this answers is not in the document]
Comment (aeyakovenko) [reply rfcs/0005-fork-selection.md]: I was indicating that it is a vote in the table, as in vote 1, 2 or 3...

File: rfcs/0005-fork-selection.md [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+| vote | vote time | lockout | lock expiration time |
+|-----:|----------:|--------:|---------------------:|
+|    7 |        11 |       2 |                   13 |
+|    1 |         1 |      16 |                   17 |
+
+The lockout for `vote 1` will not increase from 16 until the stack contains 5 votes.
+
+### Slashing and Rewards
+
+The purpose of the lockouts is to force a voter to commit opportunity cost to a specific fork.  Voters that violate the lockouts and vote for a diverging fork within the lockout should be punished.  Slashing or simply freezing the voter from rewards for a long period of time can be used as punishment.
```
[the remark this answers is not in the document]
Comment (aeyakovenko) [reply rfcs/0005-fork-selection.md]: I've seen ethereum talk about punishing misbehaving validators.

File: rfcs/0005-fork-selection.md [hunk 5]
```diff
@@ -0,0 +1,166 @@
+# Fork Selection
+ This RFC defines Solana's *Nakomoto Fork Selection* algorithm based on time locks. It satisfies the following properties:
+
+
+* A voter can eventually recover from voting on a fork that doesn't become the fork with the largest finality.
```
Comment (garious) [inline rfcs/0005-fork-selection.md]: Does finality have a size?

File: rfcs/0005-fork-selection.md [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+* If the voters share a common ancestor then they will converge to a fork containing that ancestor no matter how they are partitioned, although it may not be the latest possible ancestor at the start of the partition.
```
Comment (garious) [inline rfcs/0005-fork-selection.md]: Everything after "partitioned" can be a separate sentence. And that sentence can use some word-smithing. Is there a more precise way to describe "latest possible ancestor"? And that last word, "partition", do you mean "fork"?

File: rfcs/0005-fork-selection.md [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+* Rollback requires exponentially more time for older votes than for newer votes.
+* Voters has the freedom to set a minimum finality threshold before committing a vote to a higher lockout.  This allows each voter to make a trade-off of risk and reward.
```
Comment (garious) [inline rfcs/0005-fork-selection.md]: Voters *have* the freedom

File: rfcs/0005-fork-selection.md [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+## Time
+
+For networks like Solana, time can be the PoH hash count, which is a VDF that provides a source of time before consensus. Other networks adopting this approach would need to consider a global source of time.
+
+For Solana time uniquely identifies a specific leader for fork generation.  At any given time, only 1 leader which can be computed from the ledger itself can propose a fork.  For more details, see [fork generation](0002-consensus.md) and [leader rotation](0004-leader-rotation.md).
```
Comment (garious) [inline rfcs/0005-fork-selection.md]: missing comma after "For Solana"

Comment (garious) [reply rfcs/0005-fork-selection.md]: missing commas before and after "which can be computed for the ledger itself"

File: rfcs/0005-fork-selection.md [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+
+## Algorithm
+
+The basic idea to this approach is to stack consensus votes.  Votes in the stack must be for forks that descend from each other and for forks that are valid in the ledger they are submitted to.  Each consensus vote has a `lockout` in units of time before it can be discarded.  When a vote is added to the stack, the lockouts of all the previous votes in the stack are doubled (see [Rollback](#Rollback)).  With each new vote, a voter commits the previous votes to an ever-increasing lockout.  Since at 32 votes we can consider the system to be at `max lockout` any votes with a lockout above `1<<32` are dequeued (fifo).  Dequeuing a vote is the trigger for a reward.  If a vote expires before it is dequeued, it and all the votes above it are popped (filo) from the vote stack.  The voter needs to start rebuilding the stack from that point.
```
Comment (garious) [inline rfcs/0005-fork-selection.md]: s/fifo/FIFO/
s/filo/LIFO/

File: rfcs/0005-fork-selection.md (hunk 5 above)
Comment (aeyakovenko) [reply rfcs/0005-fork-selection.md]: @garious it is up to the node.  It could be configured that the voter will not increase a lockout past 256 unless that fork has more than 50% of the global stake.

Comment (garious) [reply rfcs/0005-fork-selection.md]: I've always measured finality in time. largest == longest? Or are you talking about the cost of rollback?

Comment (aeyakovenko) [reply rfcs/0005-fork-selection.md]: @garious Yes.  In terms of economic finality.

File: rfcs/0005-fork-selection.md [hunk 10]
```diff
@@ -0,0 +1,166 @@
+# Fork Selection
+ This RFC defines Solana's *Nakomoto Fork Selection* algorithm based on time locks. It satisfies the following properties:
+
+
+* A voter can eventually recover from voting on a fork that doesn't become the fork with the desired finality.
```
Comment (garious) [inline rfcs/0005-fork-selection.md]: It bugs me that in this first bullet, you say "desired finality." That doesn't really mean anything to me. Biggest question is, how are you defining finality? Shouldn't that be explained first? Second, what makes finality "desirable"?

File: rfcs/0005-fork-selection.md [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+* If the voters share a common ancestor then they will converge to a fork containing that ancestor no matter how they are partitioned. Although it may not be the latest possible ancestor at the start of the fork.
```
Comment (garious) [inline rfcs/0005-fork-selection.md]: The second sentence is not a sentence.

File: rfcs/0005-fork-selection.md [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+
+### Slashing and Rewards
+
+The purpose of the lockout is to force a voter to commit opportunity cost to a specific fork.  Voters that violate the lockouts and vote for a diverging fork within the lockout should be punished.  Slashing or simply freezing the voter from rewards for a long period of time can be used as punishment.
+
+Voters should be rewarded for selecting the fork that the rest of the network selected as often as possible.  This is well-aligned with generating a reward when the vote stack is full and the oldest vote needs to be dequeued.  Thus a reward should be generated for each successful dequeue.
+
+### Economic Finality
+
+Economic finality can be calculated as the total opportunity costs due to the vote lockout at a given fork.  As a votes get further and further, buried the economic finality increases because the cost of unrolling would be the total loss of all the reward from the lockouts at that fork.
```
Comment (garious) [inline rfcs/0005-fork-selection.md]: costs -> cost

Comment (garious) [reply rfcs/0005-fork-selection.md]: "a votes"

Comment (garious) [reply rfcs/0005-fork-selection.md]: Comma should be after "buried", not before.

File: rfcs/0005-fork-selection.md [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+
+### Thresholds
+
+Each voter can independently set a threshold of network commitment to a fork before that voter commits to a fork.  For example, at vote stack index 7, the lockout is 256 time units.  A voter may withhold votes and let votes 0-7 expire unless the vote at index 7 has at greater than 50% commitment in the network.  This allows each voter to independently control how much risk to commit to a fork.  Committing to a fork faster would allow the voter to earn more rewards.
```
Comment (garious) [inline rfcs/0005-fork-selection.md]: Re: "Committing to a fork faster". "faster" doesn't seem like the right word. Maybe "earlier"?

File: rfcs/0005-fork-selection.md [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+
+In both cases, the nodes in the network have several forks to pick from concurrently, even though each fork represents a different height.
+
+### Greedy Choice for Concurrent Forks
+
+When evaluating multiple forks, each node should pick the fork that will maximize economic finality, or the latest fork if all are equal.  Economic finality can be measured in terms of lockout on a fork.
+ 
+## Simulation
+
+A test library function exists for configuring networks.
```
Comment (garious) [inline rfcs/0005-fork-selection.md]: How about adding a link to the file here?  and then moving all this code documentation into that file?

File: rfcs/0005-fork-selection.md (hunk 13 above)
Comment (aeyakovenko) [reply rfcs/0005-fork-selection.md]: at a higher frequency?

File: rfcs/0005-fork-selection.md [hunk 15]
```diff
@@ -0,0 +1,99 @@
+# Fork Selection
+ This RFC defines Solana's *Nakomoto Fork Selection* algorithm based on time locks. It satisfies the following properties:
```
Comment (garious) [inline rfcs/0005-fork-selection.md]: Can you remove any references to RFC? And move this to `book/src/fork-selection.md`? And reference it in `book/src/SUMMARY.md` under "Proposed Architectural Changes"?

File: rfcs/0005-fork-selection.md [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+* A voter can eventually recover from voting on a fork that doesn't become the fork with the desired finality.
+* If the voters share a common ancestor then they will converge to a fork containing that ancestor no matter how they are partitioned, although it may not be the latest possible ancestor at the start of the fork.
+* Rollback requires exponentially more time for older votes than for newer votes.
+* Voters have the freedom to set a minimum finality threshold before committing a vote to a higher lockout.  This allows each voter to make a trade-off between risk and reward. See [finality](#economic-finality).
+
+## Time
+
+For networks like Solana, time can be the PoH hash count, which is a VDF that provides a source of time before consensus. Other networks adopting this approach would need to consider a global source of time.
+
+For Solana, time uniquely identifies a specific leader for fork generation.  At any given time only 1 leader, which can be computed from the ledger itself, can propose a fork.  For more details, see [fork generation](0002-fork-generation.md) and [leader rotation](0004-leader-rotation.md).
```
Comment (garious) [inline rfcs/0005-fork-selection.md]: After rebasing and moving this file, these links will need updating.

File: rfcs/0005-fork-selection.md [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+
+In both cases, the nodes in the network have several forks to pick from concurrently, even though each fork represents a different height.
+
+### Greedy Choice for Concurrent Forks
+
+When evaluating multiple forks, each node should pick the fork that will maximize economic finality, or the latest fork if all are equal.  Economic finality can be measured in terms of lockout on a fork.
+ 
+## Simulation
+
+Simulation can be found [here](../tests/fork-selection.rs).
```
Comment (garious) [inline rfcs/0005-fork-selection.md]: Thanks for this change, but in hindsight, the reference is actually backward. We want the code to reference the spec. Can you move the cross-reference to "tests/fork-selection.rs"?

Comment (aeyakovenko) [reply rfcs/0005-fork-selection.md]: It's kind of nice to reference code from a book.  Maybe we can move it to book/tests or book/srcs/fork-selection.rs

Comment (garious) [reply rfcs/0005-fork-selection.md]: Love this idea! Move the whole thing into `book/src` and run those assertions from Rust snippets in book content via `make test`.

File: book/src/fork-selection.md [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+
+### Slashing and Rewards
+
+The purpose of the lockout is to force a voter to commit opportunity cost to a specific fork.  Voters that violate the lockouts and vote for a diverging fork within the lockout should be punished.  Slashing or simply freezing the voter from rewards for a long period of time can be used as punishment.
+
+Voters should be rewarded for selecting the fork that the rest of the network selected as often as possible.  This is well-aligned with generating a reward when the vote stack is full and the oldest vote needs to be dequeued.  Thus a reward should be generated for each successful dequeue.
+
+### Economic Finality
+
+Economic finality can be calculated as the total opportunity cost due to the vote lockout at a given fork.  As votes get further and further buried, the economic finality increases because the cost of unrolling would be the total loss of all the reward from the lockouts at that fork.
```
Comment (garious) [inline book/src/fork-selection.md]: I think way more effort needs to go into defining this term.  I don't think of it as something that "increases." There's probably some term closely related to economic finality that increases. Maybe instead of saying "economic finality can be calculated" say something like "<yada yada yada> is called *economic finality*". How about: "a transaction is traditionally labeled *finalized* if a cluster can commit never to reverse it. Solana is an availability system and the cluster naturally breaks into partition as nodes have different views of the active set. Because of this partitioning, finalization is qualified as *economically finalized*. It means the transaction is finalized under the assumption that the fullnode had been following the correct leader. Unrolling transactions to change leaders comes at a cost - all rewards for validating those transactions. That cost is called the *unrolling cost* and can be used to determine *how finalized* a transaction really is. The higher the unrolling cost, the more finalized the transaction."

Comment (garious) [reply book/src/fork-selection.md]: Used in sentences: "A transaction, on average, is confirmed in under 1 second and is 100sol-finalized in under 3 seconds. Unrolling the transaction would cost the partition 100 sols."

Comment (garious) [reply book/src/fork-selection.md]: Incorrect usage, "The economic finality of the transaction is 100 sols."
Correct usage, "The unrolling cost of the transaction is 100 sols."
Correct usage, "Solana implements economic finality, a variation of finality that is qualified by the cost to convince a fullnode to unroll the transaction."
Correct usage, "We'll send your 1 sol transaction once the first transfer is 10sol-finalized. We'll send the 50 sol transaction after it's 100sol-finalized."

Comment (aeyakovenko) [reply book/src/fork-selection.md]: * "following correct leader" doesn't really make sense.
* network has a synchronous active set, it just doesn't have a synchronized view of the latest fork
> Used in sentences: "A transaction, on average, is confirmed in under 1 second and is 100sol-finalized in under 3 seconds. Unrolling the transaction would cost the partition 100 sols."
It would depend on how many validators voted on fork

Comment (garious) [reply book/src/fork-selection.md]: How about a counter proposal? Some definitions and a set of correct and incorrect usages? The document doesn't define it and then uses it inconsistently. Worse, you replied to one comment above with "in terms of economic finality." What in the world does that mean?

Comment (aeyakovenko) [reply book/src/fork-selection.md]:
```
Finality of *fork A* is defined as the cost in terms of time to the validators                                                                             
to confirm any other fork that does not include *fork A* as an ancestor.     
                                                                                                                                                                                                 
Economic Finality of a fork can be calculatated as the loss of all the    
rewards from *fork A* and its decendants, plus the oportunity cost of reward
due to the exponentially growing lockout. 
```

Comment (aeyakovenko) [reply book/src/fork-selection.md]: @garious We are using time as the physical bound on finality.   It looks like a combination of 1 and 3, but not exactly.  To actually cause rollback the network still has to spend real time to get to the rollback point.

Comment (garious) [reply book/src/fork-selection.md]: Please, just looking for the answer to the one question. I can't help you otherwise. Is your aim to be consistent with industry precedent or to use the same terms with different definitions?

Comment (aeyakovenko) [reply book/src/fork-selection.md]: @garious to be consistent with the industry.  Finality means something different in Casper vs Bitcoin, so it has a different meaning for us as well.  But it is something that can be measured in $.  Bitcoin it is cpu/electricity, casper it is stake, for us it is time + stake.

File: book/src/fork-selection.md [hunk 19]
```diff
@@ -1,75 +1,115 @@
...
 
 ## Algorithm
 
-The basic idea to this approach is to stack consensus votes.  Votes in the stack must be for forks that descend from each other and for forks that are valid in the ledger they are submitted to.  Each consensus vote has a `lockout` in units of time before it can be discarded.  When a vote is added to the stack, the lockouts of all the previous votes in the stack are doubled (see [Rollback](#Rollback)).  With each new vote, a voter commits the previous votes to an ever-increasing lockout.  Since at 32 votes we can consider the system to be at `max lockout` any votes with a lockout above `1<<32` are dequeued (FIFO).  Dequeuing a vote is the trigger for a reward.  If a vote expires before it is dequeued, it and all the votes above it are popped (LIFO) from the vote stack.  The voter needs to start rebuilding the stack from that point.
+The basic idea to this approach is to stack consensus votes.  Votes in the
+stack must be for forks that descend from each other and for forks that are
+valid in the ledger they are submitted to.  Each consensus vote has a `lockout`
+in units of time before it can be discarded.  When a vote is added to the
```
Comment (rob-solana) [inline book/src/fork-selection.md]: what's the smallest lockout period?  one block?

File: book/src/fork-selection.md (hunk 18 above)
Comment (garious) [reply book/src/fork-selection.md]: For reference, https://vitalik.ca/files/casper_note.html, there's no usage of finality or economic finality that includes a measurement in $. Instead, the definition of economic finality references a different term measured in $.

> This value X is called the *cryptoeconomic security margin* of the finality mechanism.

File: book/src/fork-selection.md (hunk 19 above)
Comment (aeyakovenko) [reply book/src/fork-selection.md]: @rob-solana i had 2 blocks in simulation.  Not sure, maybe 2 slots?

Comment (rob-solana) [reply book/src/fork-selection.md]: if a block is ~800ms, then max lockout is arrived at in ~25s, and lasts for ~100 years?

File: book/src/fork-selection.md [hunk 20]
```diff
@@ -0,0 +1,147 @@
+# Fork Selection
+This chapter defines Solana's *Nakomoto Fork Selection* algorithm based on time
```
Comment (garious) [inline book/src/fork-selection.md]: I don't see how this relates to Nakomoto *Block* Selection. Can you cite the description you're looking at?

File: book/src/fork-selection.md [hunk 21]
```diff
@@ -0,0 +1,147 @@
...
+locks. It satisfies the following properties:
```
Comment (garious) [inline book/src/fork-selection.md]: Of these properties, which ones does Nakomoto Block Selection satisfy and which does this architecture add?

File: book/src/fork-selection.md [hunk 22]
```diff
@@ -0,0 +1,147 @@
...
+
+
+* A voter can eventually recover from voting on a fork that doesn't become the
+  fork with the desired finality.
```
Comment (garious) [inline book/src/fork-selection.md]: I think it would help to move the definition of finality above this. This term "desired finality" still doesn't make sense to me.

File: book/src/fork-selection.md [hunk 23]
```diff
@@ -0,0 +1,147 @@
...
+
+For Solana, time uniquely identifies a specific leader for fork generation.  At
+any given time only 1 leader, which can be computed from the ledger itself, can
+propose a fork.  For more details, see [fork generation](fork-generation.md)
+and [leader rotation](leader-rotation.md).
+
+## Algorithm
+
+The basic idea to this approach is to stack consensus votes.  Votes in the
+stack must be for forks that descend from each other and for forks that are
```
Comment (garious) [inline book/src/fork-selection.md]: "descend from each other" -> "share a common ancestor"?

File: book/src/fork-selection.md [hunk 24]
```diff
@@ -0,0 +1,147 @@
...
+valid in the ledger they are submitted to.  Each consensus vote has a `lockout`
+in units of time before it can be discarded.  When a vote is added to the
```
Comment (garious) [inline book/src/fork-selection.md]: When you say "it can be discarded" what is "it" exactly? And what do you mean be "discarded"?

File: book/src/fork-selection.md [hunk 25]
```diff
@@ -0,0 +1,147 @@
...
+stack, the lockouts of all the previous votes in the stack are doubled (see
+[Rollback](#Rollback)).  With each new vote, a voter commits the previous votes
+to an ever-increasing lockout.  Since at 32 votes we can consider the system to
```
Comment (garious) [inline book/src/fork-selection.md]: Why 32?

File: book/src/fork-selection.md [hunk 26]
```diff
@@ -0,0 +1,147 @@
...
+|-----:|----------:|--------:|---------------------:| |    7 |        11 |
+2 |                   13 | |    1 |         1 |      16 |                   17
+|
+
+The lockout for vote 1 will not increase from 16 until the stack contains 5
+votes.
+
+### Slashing and Rewards
+
+The purpose of the lockout is to force a voter to commit opportunity cost to a
```
Comment (garious) [inline book/src/fork-selection.md]: It would have been helpful to read about this purpose much earlier.

File: book/src/fork-selection.md (hunk 20 above)
Comment (aeyakovenko) [reply book/src/fork-selection.md]: https://www.reddit.com/r/btc/comments/4r0mfa/longest_chain_or_most_work/

nodes prefer the chain with the most work, and not the "longest" one.  our chain selection rule is to pick the chain with the most lockout, and not the longest one.

File: book/src/fork-selection.md (hunk 18 above)
Comment (aeyakovenko) [reply book/src/fork-selection.md]: @garious what do you want to call this then?  "rollback difficulty"

File: book/src/fork-selection.md (hunk 25 above)
Comment (aeyakovenko) [reply book/src/fork-selection.md]: 2^32 is sufficiently large

File: book/src/fork-selection.md (hunk 23 above)
Comment (aeyakovenko) [reply book/src/fork-selection.md]: @garious each vote on the stack is an ancestor of the vote above it.

File: book/src/fork-selection.md (hunk 18 above)
Comment (garious) [reply book/src/fork-selection.md]: I suggested "cost of rollback" earlier.

File: book/src/fork-selection.md (hunk 19 above)
Comment (aeyakovenko) [reply book/src/fork-selection.md]: @rob-solana we can make the curve whatever we want.  `C + (2^N)/K`

File: book/src/fork-selection.md (hunk 18 above)
Comment (aeyakovenko) [reply book/src/fork-selection.md]: @garious The **Economic Finality** of *fork A* can be calculated as the loss of all the                                                                                                                                                                                                                                                                                               rewards from rollback of *fork A* and its descendants, plus the opportunity cost of reward                                                                                                                                                                                                                                                                                                
due to the exponentially growing lockout of the votes that have confirmed *fork A*.

File: book/src/fork-selection.md (hunk 21 above)
Comment (aeyakovenko) [reply book/src/fork-selection.md]:
```
* Rollback requires exponentially more time for older votes than for newer
  votes.
```
in PoW chain rollback requires exponentially more electricity to overcome the current chain.

File: book/src/fork-selection.md (hunk 22 above)
Comment (aeyakovenko) [reply book/src/fork-selection.md]: how about *network confirmation threshold*?

File: book/src/fork-selection.md (hunk 24 above)
Comment (aeyakovenko) [reply book/src/fork-selection.md]: Each vote in the stack is a confirmation of a fork.  Each confirmed fork is an ancestor of the  fork above it.  Each consensus vote has a `lockout` in units of time before the voter can submit a vote that does not contain the confirmed fork as an ancestor.

File: book/src/fork-selection.md (hunk 26 above)
Comment (aeyakovenko) [reply book/src/fork-selection.md]: you want rearrange this around after we finalize the terminology?

File: rfcs/0005-fork-selection.md [hunk 27]
```diff
@@ -0,0 +1,166 @@
+# Fork Selection
+ This RFC defines Solana's *Nakomoto Fork Selection* algorithm based on time locks. It satisfies the following properties:
+
+
+* A voter can eventually recover from voting on a fork that never sees majority votes.
+* If the voters share a common ancestor then they will converge to a fork containing that ancestor no matter how they are partitioned.  Although it may not be the latest possible ancestor at the start of the partition.
```
[the remark this answers is not in the document]
Comment (aeyakovenko) [reply rfcs/0005-fork-selection.md]: fixed!

File: rfcs/0005-fork-selection.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (garious) [reply rfcs/0005-fork-selection.md]: They shouldn't be using language that causes honest stakers concern either. Marking this one resolved since it's not too important.

File: book/src/fork-selection.md (hunk 21 above)
Comment (garious) [reply book/src/fork-selection.md]: So maybe add a section that says how this algorithm compares to Nakomoto's? Or rephrase this to say it's "similar in spirit because like Nakomoto Block Selection, rollback requires exponentially more *something* for older *something*. Looks like a tricky comparison to me. Moving it into a separate section should minimize the wtfs-per-sec as someone familiar with Nakomoto Block Selection reads about time and votes.

Comment (garious) [reply book/src/fork-selection.md]: Or said another way, it'd be good to see an uncontroversial first sentence in this proposal. :)

File: rfcs/0005-fork-selection.md (hunk 27 above)
[the remark this answers is not in the document]
Comment (garious) [reply rfcs/0005-fork-selection.md]: Indeed the second sentence is now a sentence. New problem: "converged ancestor" is nonsense.

File: book/src/fork-selection.md (hunk 20 above)
Comment (garious) [reply book/src/fork-selection.md]: Seems I wasn't specific enough with my request. Can you cite the description in the document?

File: tests/fork-selection.rs [hunk 28]
```diff
@@ -0,0 +1,598 @@
+//! Fork Selection Simulation
```
Comment (garious) [inline tests/fork-selection.rs]: What happened to the plan to move this to `book/src`, running assertions from markdown above, which would be run via `make -C book test`?

**Comment by aeyakovenko** [comment]
@garious death by a thousand cuts!

File: book/src/fork-selection.md (hunk 20 above)
Comment (aeyakovenko) [reply book/src/fork-selection.md]: just leaving this as a note:

Honest generators only build onto a block (by referencing it in blocks they create) if it is the latest block in the longest valid chain. "Length" is calculated as total combined difficulty of that chain, not number of blocks, though this distinction is only important in the context of a few potential attacks. A chain is valid if all of the blocks and transactions within it are valid, and only if it starts with the genesis block. (source https://en.bitcoin.it/wiki/Block_chain)

File: tests/fork-selection.rs (hunk 28 above)
Comment (aeyakovenko) [reply tests/fork-selection.rs]: lets do that in another PR, it wasn't clear what the Makefile should look like to run the cargo tests as well.
